physics/GFS_rrtmgp_cloud_mp.F90

```diff
@@ -0,0 +1,883 @@
+  !
+  ! - The liquid and ice cloud effective particle sizes are assigned reference values*.
+  !   *TODO* Find references, include DOIs, parameterize magic numbers, etc...
+  !
+  ! ######################################################################################
+  subroutine cloud_mp_MYNN(nCol, nLev, lsmask, t_lay, p_lev, p_lay, qs_lay, relhum,      &
+       qc_mynn, qi_mynn, con_ttp, con_g, cld_pbl_lwp, cld_pbl_reliq, cld_pbl_iwp,     &
+       cld_pbl_reice, cld_pbl_frac)
+    implicit none
+
+    ! Inputs
+    integer, intent(in)    :: &
+         nCol,          & ! Number of horizontal grid points
+         nLev             ! Number of vertical layers
+    real(kind_phys), dimension(:), intent(in) :: &
+         lsmask           ! Land/Sea mask
+    real(kind_phys), intent(in) :: &
+         con_g,         & ! Physical constant: gravitational constant 
+         con_ttp          ! Triple point temperature of water (K)
+    real(kind_phys), dimension(:,:),intent(in) :: &
+         t_lay,         & ! Temperature at layer centers (K)
+         p_lev,         & ! Pressure at layer interfaces (Pa)
+         p_lay,         & !
+         qs_lay,        & !
+         relhum,        & !
+         qc_mynn,       & ! Liquid cloud mixing-ratio (MYNN PBL cloud)
+         qi_mynn,       & ! Ice cloud mixing-ratio (MYNN PBL cloud)
+         cld_pbl_frac    ! Cloud-fraction (MYNN PBL cloud)
+    ! Outputs
+    real(kind_phys), dimension(:,:),intent(inout) :: &
+         cld_pbl_lwp,   & ! Convective cloud liquid water path
+         cld_pbl_reliq, & ! Convective cloud liquid effective radius
+         cld_pbl_iwp,   & ! Convective cloud ice water path
+         cld_pbl_reice    ! Convective cloud ice effecive radius
+    
+    ! Local
+    integer :: iCol, iLay
+    real(kind_phys) :: tem1, qc, qi, deltaP
+
+    tem1 = 1.0e5/con_g
+    do iLay = 1, nLev
+       do iCol = 1, nCol
+          if (cld_pbl_frac(iCol,iLay) > cld_limit_lower) then
+             ! Cloud mixing-ratios (DJS asks: Why is this done?)
+             qc = qc_mynn(iCol,iLay)*cld_pbl_frac(iCol,iLay)
+             qi = qi_mynn(iCol,iLay)*cld_pbl_frac(iCol,iLay)
+
+             ! LWP/IWP
+             deltaP = abs(p_lev(iCol,iLay+1)-p_lev(iCol,iLay))
+             cld_pbl_lwp(iCol,iLay) = max(0., qc * tem1 * deltaP)
+             cld_pbl_iwp(iCol,iLay) = max(0., qi * tem1 * deltaP)
+
+             ! Particle sizes
+             if (nint(lsmask(iCol)) == 1) then
+                if(qc > 1.E-8) cld_pbl_reliq(iCol,iLay) = 5.4
+             else
+                ! Cloud water (microns), from Miles et al.
+                if(qc > 1.E-8) cld_pbl_reliq(iCol,iLay) = 9.6
+             endif
+             ! Cloud ice (microns), from Mishra et al. (2014, JGR Atmos, fig 6b) 
+             if(qi > 1.E-8) cld_pbl_reice(iCol,iLay) = max(173.45 + 2.14*(t_lay(iCol,iLay)-273.15), 20.)
+          endif
+       enddo
+    enddo
+  end subroutine cloud_mp_MYNN
```

Comment:
We haven't started testing RRTMGP yet, but I'm sure this code/module will need to be revisited. This seems to be a partial duplication of SGSCloud_RadPre and I'm not certain yet about the order of calls. This concern is not just for the MYNN but also for GF. We need to have this code in one module only.

Reply:
Same comment here: It will only cause problems when we start testing with RRTMGP. It looks like the organization/improvement changes they made to the radiation code were just a first step.